chainstate/src/detail/mod.rs

```diff
@@ -430,7 +435,17 @@ impl<S: BlockchainStorage, V: TransactionVerificationStrategy> Chainstate<S, V>
         block: WithId<Block>,
     ) -> Result<WithId<Block>, BlockError> {
         let chainstate_ref = self.make_db_tx_ro().map_err(BlockError::from)?;
-        chainstate_ref.check_block(&block).log_err()?;
+
+        let best_block_id = chainstate_ref
+            .get_best_block_id()
+            .map_err(BlockError::BestBlockLoadError)
+            .log_err()?;
+        let best_block_height = chainstate_ref
+            .get_block_height_in_main_chain(&best_block_id)
+            .map_err(BlockError::BestBlockLoadError)
+            .log_err()?
+            .expect("best block height doesn't exist");
+        chainstate_ref.check_block(&block, best_block_height.next_height()).log_err()?;
```

Comment:
`best_block_height.next_height()` here is wrong. This implies that the current best block is always the previous block of the block you're checking. 

You should get the block height of the previous block, then do `next_height()` on that. Not the best block.

Reply:
A good question is: How didn't the tests catch this mistake?!

Reply:
> A good question is: How didn't the tests catch this mistake?!

It requires a very specific test to catch. But at the moment we support only producing PoS blocks from the tip. I'll add the test but leave it ignored for now.